woltka/workflow.py

```diff
@@ -0,0 +1,590 @@
+        be "none" to omit classification (simply report subject IDs) or "free"
+        to perform free-rank classification (LCA of subjects regardless of rank
+        will be reported).
+    rank2dir : dict, otional
+        Write classification map per rank to directory.
+
+    above : bool, optional
+        Allow assigning to a classification unit higher than given rank.
+    major : int, optional
+        Perform majority-rule assignment based on this percentage threshold.
+        Range: [51, 99].
+    ambig : bool, optional
+        Allow one sequence to be assigned to multiple classification units at
+        the same rank. The profile will be normalized by the number of matches
+        per query sequence.
+    subok : bool, optional
+        Allow directly reporting subject ID(s) if a sequence cannot be assigned
+        to any higher classification unit.
+    deidx : bool, optional
+        Strip "underscore index" suffixes from subject IDs.
+
+    lines : int, optional
+        Number of lines to read from alignment file per chunk.
+
+    Returns
+    -------
+    dict of dict
+        Per-rank profiles generated from classification.
+    """
+    data = {x: {} for x in ranks}
+
+    # assignment parameters
+    kwargs = {'tree':   tree, 'rankd': rankd,  'root':  root, 'above': above,
+              'major': major, 'ambig': ambig, 'subok': subok, 'named': named,
+              'rank2dir': rank2dir}
+
+    # parse input maps and generate profile
+    for fp in sorted(files):
+        n = 0
+        click.echo(f'Parsing alignment file {basename(fp)} ', nl=False)
+
+        # read alignment file into query-subject(s) map
+        with readzip(fp) as fh:
+
+            # parse alignment file by chunk
+            for rmap in parse_align_file(fh, proc, input_fmt, lines):
+                click.echo('.', nl=False)
+                n += len(rmap)
+
+                # reshape read map
+                rmap = reshape_rmap(rmap, deidx, demux, samples, files, fp)
+
+                # assign reads at each rank
+                for sample, map_ in rmap.items():
+                    for rank in ranks:
+
+                        # call assignment workflow
+                        assign_rmap(map_, data, rank, sample, **kwargs)
+
+        click.echo(' Done.')
+        click.echo(f'Query sequences: {n}.')
+
+    click.echo('Task completed.')
+    return data
+
+
+def parse_samples(path_:  str,
+                  ext:    str = None,
+                  ids:   list = None,
+                  demux: bool = None) -> (list, list or dict, bool):
+    """Determine sample IDs, aligment files, and multiplex status.
+
+    Parameters
+    ----------
+    path_ : str
+        Path to a file or a directory.
+    ext : str, optional
+        Filename extension.
+    ids : iterable of str, optional
+        List of sample IDs.
+    demux : bool, optional
+        Whether perform demultiplexing.
```

Comment:
specify default value and behavior with default value

Reply:
For this particular case, specifying default is bit tricky, because when it is `None`, the program will determine whether it should be `True` or `False` based on the input file(s). This comes back to a previous question on whether behavior should be explicit or implicit (see above). Btw, I think specifying default is not mandatory in numpy-style docstring, especial when default is `None`.

Reply:
Maybe specifying default is not _mandatory_. But you should very clearly state what happens when no parameter is specified.